### PR TITLE
[ONNX] Warning when using __len__ to calculate tensor shape (#55151)

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57830 [ONNX] Support index_add_ function. (#56867)
* #57829 [ONNX] Handling incorrect format for example_outputs  (#55802)
* #57605 [ONNX] Fix split export in opset13 (#56277)
* #57604 [ONNX] Handle mixed mask, index input for index_put
* #57603 [ONNX] ListUnpack on dynamic tensor list (#56592)
* #57602 [ONNX] Handle NoneType in Assign Output Shapes (#54623)
* #57601 [ONNX] Process const folding progressively when converts to ONNX (#54569)
* #57600 [ONNX] Support registering custom export for prim::PythonOp from torch.autograd.Function (#55630)
* #57599 Add support to to(device) op. (#56857)
* #57598 onnx.symbolic_helper.parse_args: document and clean up (#56956)
* #57597 [ONNX] Special post process for onnx::Cast and onnx::ConstantOfShape shape type inference (#55962)
* #57596 Add a new operator for fill_() function. (#56859)
* **#57595 [ONNX] Warning when using __len__ to calculate tensor shape (#55151)**
* #57594 [ONNX] Support .item() export & NumberType to tensor conversion (#55697)

Difference in traced graph and outputs, when using len(tensor) as compared to tensor.shape[0]

An example model is (with tensor.shape):
```
# Test len fix with variable inputs
import torch
import onnxruntime

class Model(torch.nn.Module):
    def forward(self, x):
        return x.size(1) + x.shape[0]


# Call export
dummy_x = torch.randn(3, 5)
model = Model()

import io
onnx_io = io.BytesIO()
torch.onnx.export(model, (dummy_x,), onnx_io,
                  input_names=['input'],
                  dynamic_axes={'input': {0:'h'}},
                  verbose=True)

# Call onnxruntime runtime and compare outputs on dynamic inputs
ort_session = onnxruntime.InferenceSession(onnx_io.getvalue())

x = torch.randn(2, 5)
print(model(x))
print(ort_session.run(None, {ort_session.get_inputs()[0].name: x.numpy()}))
```
The output graph is as follows:
```
graph(%input : Float(*, 5, strides=[5, 1], requires_grad=0, device=cpu)):
  %1 : Long(2, strides=[1], device=cpu) = onnx::Shape(%input)
  %2 : Long(device=cpu) = onnx::Constant[value={1}]()
  %3 : Long(device=cpu) = onnx::Gather[axis=0](%1, %2) # test/onnx/test_m.py:9:0
  %4 : Long(2, strides=[1], device=cpu) = onnx::Shape(%input)
  %5 : Long(device=cpu) = onnx::Constant[value={0}]()
  %6 : Long(device=cpu) = onnx::Gather[axis=0](%4, %5) # test/onnx/test_m.py:9:0
  %7 : Long(requires_grad=0, device=cpu) = onnx::Add(%3, %6) # test/onnx/test_m.py:9:0
  return (%7)
```
Torch output: 7
ORT output: 7

Now replacing tensor.shape[0] with len(tensor), the graph looks like:
```
graph(%input : Float(*, 5, strides=[5, 1], requires_grad=0, device=cpu)):
  %1 : Long(2, strides=[1], device=cpu) = onnx::Shape(%input)
  %2 : Long(device=cpu) = onnx::Constant[value={1}]()
  %3 : Long(device=cpu) = onnx::Gather[axis=0](%1, %2) # test/onnx/test_m.py:9:0
  %4 : Long(requires_grad=0, device=cpu) = onnx::Constant[value={3}]()
  %5 : Long(requires_grad=0, device=cpu) = onnx::Add(%3, %4)
  return (%5)
```
Torch output: 7
ORT output: 8

In the case with __len__, %4 is traced as a constant

**Workaround to avoid the mismatch when using len to get tensor.shape**

Add the following pattern around _export call
```
    import builtins
    len_backup = builtins.len
    builtins.len = lambda x : x.__len__()

    # Call export
    _export(model, args, .....

    builtins.len = len_backup

```

Co-authored-by: shubhambhokare1 <shubhambhokare1@gmail.com>

Differential Revision: [D28393526](https://our.internmc.facebook.com/intern/diff/D28393526)

### CODE REVIEW

**Comment by SplitInfinity** [comment]
@SplitInfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28393526).
